issue-27 allow code to randomize the returned ip addresses

mahtin opened the pull request:
Related to #27. This code randomize the returned ip addresses so that the load is shared around the pool.

cf-natali commented on ntplib.py:
```diff
@@ -298,7 +299,8 @@ def request(self, host, version=2, port="ntp", timeout=5, address_family=socket.
 
         # lookup server address
-        addrinfo = socket.getaddrinfo(host, port, address_family)[0]
+        addrinfo_a = socket.getaddrinfo(host, port, address_family)
```
Simpler:
```
addrinfo = random.choice(socket.getaddrinfo(host, port, address_family))
```

mahtin replied:
Yes - much better! (I'll update my pull request)

cf-natali reviewed (commented):
Besides the attached comment, the problem with this change is that it can break things if some DNS records can't be used.

For example `2.pool.ntp.org` returns some AAAA records for IPv6, which currently the code would ignore, but with that change, if we randomly use an IPv6 address and the host/network doesn't support it, it would randomly break.

Example:
```
cf@thinkpad:~/src/ntplib$ nslookup 2.pool.ntp.org
Server:         1.1.1.1
Address:        1.1.1.1#53

Non-authoritative answer:
Name:   2.pool.ntp.org
Address: 213.5.132.231
Name:   2.pool.ntp.org
Address: 178.79.188.22
Name:   2.pool.ntp.org
Address: 193.150.34.2
Name:   2.pool.ntp.org
Address: 103.219.22.112
Name:   2.pool.ntp.org
Address: 2a01:7e00::f03c:91ff:fe96:f129
Name:   2.pool.ntp.org
Address: 2a03:b0c0:1:d0::25c:1
Name:   2.pool.ntp.org
Address: 2a0b:9b00:463::123
Name:   2.pool.ntp.org
Address: 2a00:2381:19c6::200
```

```
cf@thinkpad:~/src/ntplib$ git diff
diff --git a/ntplib.py b/ntplib.py
index 77ae7b7..b10745e 100644
--- a/ntplib.py
+++ b/ntplib.py
@@ -29,6 +29,7 @@ functions.
 
 
 import datetime
+import random
 import socket
 import struct
 import time
@@ -298,7 +299,7 @@ class NTPClient(object):
         """
 
         # lookup server address
-        addrinfo = socket.getaddrinfo(host, port, address_family)[0]
+        addrinfo = random.choice(socket.getaddrinfo(host, port, address_family))
         family, sockaddr = addrinfo[0], addrinfo[4]
 
         # create the socket
```

```
cf@thinkpad:~/src/ntplib$ python3 -c "from ntplib import NTPClient; c = NTPClient(); c.request('2.pool.ntp.org')"
cf@thinkpad:~/src/ntplib$ python3 -c "from ntplib import NTPClient; c = NTPClient(); c.request('2.pool.ntp.org')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/cf/src/ntplib/ntplib.py", line 319, in request
    sock.sendto(query_packet.to_data(), sockaddr)
OSError: [Errno 101] Network is unreachable
cf@thinkpad:~/src/ntplib$ 
```

See `ENETUNREACH (errno=101)` when we pick an Ipv6 address at random.

mahtin commented:
Good point (and I so happen to have some IPv6-only operational knowledge). I'll update my pull request; however, there's always going to a case in any internet-related application where IPv6 only hosts are referenced at a time where a user is siting on a suboptimal IPv4-only connection (think "mythic beasts" in the UK and their IPv6-only hosting option). You can't code away from that situation. Looping thru the list randomly till success could be the best resolution.

mahtin commented:
BTW: Thank you for using the DNS 1.1.1.1 resolver (from my previous job and current car) 😁

![DNS1111](https://user-images.githubusercontent.com/4521085/191840666-c53cae1d-ff1d-4f5d-bdb5-7ce01ed3a98f.jpeg)

mahtin commented:
Ok, so this was interesting (and I had to drop v6 to test this - yikes!) I'm happy with the updated PR.

I added a hunk of extra checking into the logic plus I added code to limit the protocol to UDP at all points where it mattered. Now the `getaddrinfo()` call only asks for `SOCK_DGRAM` entries. Previously, the `getaddrinfo()` call would respond with both `SOCK_DGRAM` and `SOCK_STREAM` entries for every host address. Those duplicated addressed was subsequently ignored because of the previous `[0]` code (which was fine - the previous logic still worked). The `[0]` code is now replaced by the randomizing code. The array of addresses is returned and looped-over after that randomization and also limited to `_MAX_HOSTS_TO_TRY` (8) entries just to make sure there's no run-away cpu usage by the `random.sample()` call.

One large change that could be further discussed (however, I did test it and I'm happy with this PR) is the timeout code. As the code now loops over `n` addresses. The timeout logic is now based on `timeout/n` per loop. The total timeout is still the same if everything fails; however, you could say this logic is simplistic at-best. I won't disagree. I'll leave any updates to that for a later PR.

Finally - I would like to say that this code is really useful for a project I'm working on. Thank you.

cf-natali commented on ntplib.py:
```diff
@@ -280,6 +281,8 @@ def dest_time(self):
     """NTP client session."""
 
+    _MAX_HOSTS_TO_TRY = 8 # Normally pool responses are 4 hosts and hence a max of 8 v4+v6 addresses; but other hosts could be higher
+
```
I'm not a fan of this: I don't believe a DNS response could have enough entries for it to be a problem in practice.

mahtin replied:
This was defined out of an abundance of caution such that should a DNS response be massive, the `random.sample()` call doesn't work on and return a massive dataset. To be honest, if `_MAX_HOSTS_TO_TRY` was 1000, the code would still be near-instant. I'm happy to up that value above `8`; however, keep in mind that the current code operates with `[0]` and hence even a max value of 1 provides a similar logic and results. Going back and looking at the NTP pool responses, I only see eight address values being returned.

cf-natali commented on ntplib.py:
```diff
@@ -297,37 +300,62 @@ def request(self, host, version=2, port="ntp", timeout=5, address_family=socket.
+        # if a specific address does not work, no worries, we just continue to the next
+        # for sanity reasons (i.e. limiting random sample cpu usage); we limit the tries to _MAX_HOSTS_TO_TRY addresses
+        for addrinfo in random.sample(addrinfo_a, k=min(NTPClient._MAX_HOSTS_TO_TRY, len(addrinfo_a))):
+            if addrinfo[1] != socket.SOCK_DGRAM or addrinfo[2] != socket.IPPROTO_UDP:
```
I don't think this should happen if we pass `SOCK_DGRAM` to `getaddrinfo` above, so might as well remove.

mahtin replied:
This is me defaulting to "belt and suspenders"[1] style coding. In this case it's all about documenting and coding the correct path thru the code. I agree it's not 100% needed; however, it does force the code into never hitting a failure mode.

[1] Definition: "Involving or employing multiple methods or procedures to achieve a desired result especially out of caution or fear of failure"

cf-natali commented on ntplib.py:
```diff
@@ -297,37 +300,62 @@ def request(self, host, version=2, port="ntp", timeout=5, address_family=socket.
+            # create the socket for UDP datagram
+            try:
+                sock = socket.socket(family, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
+            except OSError as e:
```
Since it should not happen, I'd rather let the error propagate, to make debugging easier if it does happen.

mahtin replied:
I think I have the same answer as above; however, with all this, it's your code. Feel free to delete.

That said, if you are on a machine with v6 not complied in to the kernel, I think this case can happen - and hence you want to `continue` and use the next returned address.  (I can't test this as I only have v6 enabled kernels). I'm mainly going after the randomization logic which then dictated the looping (once you realize the v4/v6 issue). The use of `SOCK_DGRAM` in the `socket.getaddrinfo()` call propagates downwards into the returned values.

cf-natali commented on ntplib.py:
```diff
@@ -297,37 +300,62 @@ def request(self, host, version=2, port="ntp", timeout=5, address_family=socket.
+
+            try:
+                # The timeout logic changes. We have N hosts to check, so we have to sum up the timeouts
+                sock.settimeout(timeout/min(NTPClient._MAX_HOSTS_TO_TRY, len(addrinfo_a)))
```
I think a better way to do that would be to measure the timeout from the start of the function, i.e.:
```
deadline = time.time() + timeout
for addr in getaddrinfo():
    [...]
    sock.settimeout(max(0, deadline - time.time()))
```

This way the timeout is accurate no matter how many address we try, and whether they succeed or not.

mahtin replied:
Correct. I like that. I can PR that. I think it needs an additional `if deadline - time.time() < 0` check followed by a `break` otherwise you could send a negative timeout to `socket.settimeout`[1].

[1] https://docs.python.org/3/library/socket.html?highlight=socket%20settimeout#socket.socket.settimeout "The value argument can be a nonnegative floating point number expressing seconds, or None."

mahtin replied:
OK. This logic would work...
```
$ python
...
>>> import socket
>>> sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
>>> sock.settimeout(-1.0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Timeout value out of range
>>> 
```
That can handle the negative timeout.

mahtin replied:
BTW: Yes, I see the `max(0,...)` however, if you pass 0 you would not get what you expect[1] as non-blocking mode would expect the response packet to already be there BEFORE the call to `sock.recvfrom()` which is unlikely. The code would loop needlessly sending NTP requests and ignoring the responses from past NTO requests. 

[1] If zero is given, the socket is put in non-blocking mode.

mahtin replied:
Hear me out for a second. This new timeout code `deadline - time.time()` means that if the first returned address fails (with a timeout) then it's eaten up all the time and no other addresses are left to try because the timeout has been exhausted. There's no easy fix for this, as a simple backoff algorithm (lets say: dividing the remaining time by `2`) would simply bias the first returned address as having the lions-share of the timeout. I'll leave this as a to-do issue; as, once again, even in this case, it's emulating the existing `[0]` code.

cf-natali commented:
Left several comments.

One thing I'm wondering - is this actually a problem in practice?
No one ever complained about this, and I can't imagine that simple NTP queries would need to be load-balanced for reasonable use cases.

Why exactly are you looking at this?

mahtin commented:
> One thing I'm wondering - is this actually a problem in practice?

So the answer is yes and no. Technically, the NTP pool produces random results; however, as that can get cached in a local'ish resolver, there's the possibility that subsequent calls would hit the same server.

The present code does not include a loop around the returned values - that alone is work fixing. 

> No one ever complained about this, and I can't imagine that simple NTP queries would need to be load-balanced for reasonable use cases.

I agree; however, once I read over the code I felt compelled to add the randomization and then add the looping code. Would a majority of users hit this issue? No. Could someone at some point hit this at some point? Yes. Absolutely. 

> Why exactly are you looking at this?

Blame OCD if you like. :) I needed an NTP package for a new project I'm working on (satellite tracking) and while reading over the code, I saw your logic.

Sitting in the back of my brain is this issue from nearly two decades ago. https://pages.cs.wisc.edu/~plonka/netgear-sntp/ and that has been an issue that I have kept in mind while building real world Internet coding.

Hence the PR. 

I'll answer your comments in the code next.

cf-natali commented:
> I agree; however, once I read over the code I felt compelled to add the randomization and then add the looping code. Would a majority of users hit this issue? No. Could someone at some point hit this at some point? Yes. Absolutely.

I understand, however I feel a bit nervous potentially breaking code which is currently perfectly working for something which in over 10 years has never been an issue - "if it ain't broke, don't fix it".
The benefit is very small, and the downside is quite large, not even mentioning the extra complexity.

mahtin commented:
> The benefit is very small, and the downside is quite large, not even mentioning the extra complexity.

As I said, it's your code.

I could go on and on and on about entropy in codebases and why that (just like the universe) is inevitable. However, I think if you test this code (I did), you'll see it has value. I'm just bringing some of my read world experience to the table.

cf-natali commented:
Thanks, I appreciate the feedback, however as mentioned this code has worked without issue for 13 years, so I'm going to keep as-is until/if it's ever an issue.

Cheers,
